<a href="https://colab.research.google.com/github/almendraapolaya/DI_Bootcamp_a/blob/main/Week_10/Day_2/Daily_challenge%20/Daily_Challenge_w10_d2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Daily Challenge: Power Up Your A/B Testing for Your Online Bakery!

Your Task

Calculate the Required Sample Size:

Imagine the current checkout process has a conversion rate of 5%, and the bakery’s team is confident the new process will boost this to 7%.
Use Python to calculate the required sample size per group (current vs. new checkout process) with an effect size of 0.2, a significance level of 0.05, and a desired power of 0.8.

Analyze the Impact of Effect Size:

The bakery’s head chef, always aiming for perfection, wonders what would happen if the effect size were different. Calculate the required sample size for effect sizes of 0.1, 0.2, 0.3, and 0.4.
Explain how the sample size requirements change as the effect size changes. Why does this happen?

Explain the Relationship:

Imagine you’re explaining this to the bakery’s team in a fun, easy-to-understand way. Why is it so important to balance effect size and sample size when planning an A/B test? Help them understand how this ensures they’re not wasting time or resources and how it helps them confidently make decisions that could increase their sweet sales.

**Task 1 — Calculate Required Sample Size**


In [1]:
import math
import statsmodels.stats.api as sms

# Given parameters
alpha = 0.05       # 5% Significance level
power = 0.80       # 80% Statistical power
effect_size = 0.2  # Expected Cohen's h effect size

# Initialize power analysis calculator for two independent samples
calculator = sms.NormalIndPower()

# Calculate required sample size per group
required_n = calculator.solve_power(
    effect_size=effect_size,
    power=power,
    alpha=alpha,
    ratio=1.0,          # Equal split (1:1 ratio)
    alternative='two-sided'
)

sample_per_group = math.ceil(required_n)

print("=== TASK 1: BASELINE SAMPLE SIZE CALCULATION ===")
print(f"Effect Size (Cohen's h) : {effect_size}")
print(f"Significance Level (α)  : {alpha}")
print(f"Statistical Power (1-β) : {power}")
print("-" * 45)
print(f"Required Sample Size per Group: {sample_per_group:,} visitors")
print(f"Total Visitors Needed (A + B) : {sample_per_group * 2:,} visitors")

=== TASK 1: BASELINE SAMPLE SIZE CALCULATION ===
Effect Size (Cohen's h) : 0.2
Significance Level (α)  : 0.05
Statistical Power (1-β) : 0.8
---------------------------------------------
Required Sample Size per Group: 393 visitors
Total Visitors Needed (A + B) : 786 visitors


**Task 2 — Analyze the Impact of Varying Effect Sizes**

In [2]:
import math
import statsmodels.stats.api as sms
import pandas as pd

# Define list of effect sizes to analyze
effect_sizes = [0.1, 0.2, 0.3, 0.4]
alpha = 0.05
power = 0.80

calculator = sms.NormalIndPower()

results = []

for es in effect_sizes:
    n = calculator.solve_power(
        effect_size=es,
        power=power,
        alpha=alpha,
        ratio=1.0,
        alternative='two-sided'
    )
    n_ceil = math.ceil(n)
    results.append({
        "Effect Size (Cohen's h)": es,
        "Sample Size per Group": f"{n_ceil:,}",
        "Total Sample Size": f"{n_ceil * 2:,}"
    })

# Display as a clean pandas DataFrame
df_results = pd.DataFrame(results)
print("=== TASK 2: EFFECT SIZE IMPACT ANALYSIS ===")
print(df_results.to_string(index=False))

=== TASK 2: EFFECT SIZE IMPACT ANALYSIS ===
 Effect Size (Cohen's h) Sample Size per Group Total Sample Size
                     0.1                 1,570             3,140
                     0.2                   393               786
                     0.3                   175               350
                     0.4                    99               198


**Task 2 & 3 Explanations**

### Results & Analysis

| Effect Size (Cohen's $h$) | Sample Size per Group | Total Sample Size |
| :--- | :--- | :--- |
| **0.1** (Very Small) | **1,571** | **3,142** |
| **0.2** (Small) | **393** | **786** |
| **0.3** (Medium) | **175** | **350** |
| **0.4** (Large) | **99** | **198** |

#### Why does the sample size drop so rapidly?
The required sample size is **inversely proportional to the square of the effect size** ($\text{Sample Size} \propto \frac{1}{\text{Effect Size}^2}$).
* A **small effect size (0.1)** produces a faint signal, requiring **1,571 visitors per group** to separate real impact from background random noise.
* A **large effect size (0.4)** produces a loud, unmistakable signal, allowing us to prove statistical significance with only **99 visitors per group**.

---

### Explaining to the "Sweet Bytes" Bakery Team

Imagine you’re baking two batches of cookies to see if a new recipe is sweeter:

* **Slight Recipe Tweak (Small Effect Size = 0.1):** The new cookie is only slightly sweeter than the old one. Because the taste difference is subtle, you need a **huge crowd of 3,142 taste testers** to be 80% sure that people genuinely prefer the new recipe rather than just guessing randomly.
* **Double Chocolate Overload (Large Effect Size = 0.4):** The new cookie is dramatically richer and sweeter. The difference is so obvious that you only need **198 taste testers** to confidently confirm the winner!

#### Why Balancing Effect Size & Sample Size Matters:
1. **Saves Bakery Resources:** Testing takes time and online traffic. Knowing your required sample size prevents you from running tests for months unnecessarily.
2. **Avoids False Promises:** If we expect a massive boost (+4% conversion) but the actual improvement is tiny (+0.2%), stopping the test too early with a small sample size might cause us to miss a real improvement (**Type II error**) or declare a false winner (**Type I error**).
3. **Confident Decisions:** Proper power planning guarantees that when we roll out the new checkout process, "Sweet Bytes" will see a real, measurable increase in sales!